src/drive/web/modules/viewer/FilesViewer.jsx

```diff
@@ -35,140 +42,168 @@ export const FilesViewerLoading = () => (
+    nextFile => {
+      if (onChange) {
+        onChange(nextFile.id)
+      }
+    },
+    [onChange]
+  )
+
+  const getCurrentIndex = useCallback(
+    () => files.findIndex(f => f.id === fileId),
+    [files, fileId]
+  )
 
   // If we can't find the file in the loaded files, that's probably because the user
   // is trying to open a direct link to a file that wasn't in the first 50 files of
   // the containing folder (it comes from a fetchMore...) ; we load the file attributes
   // directly as a contingency measure
-  async fetchFileIfNecessary() {
-    if (this.getCurrentIndex() !== -1) return
-    if (this.state.currentFile && this._mounted) {
-      this.setState({ currentFile: null })
-    }
-
-    const { fileId, client } = this.props
-    try {
-      const { data } = await client.query(Q('io.cozy.files').getById(fileId))
-      this.setState({
-        currentFile: data
-      })
-    } catch (e) {
-      logger.warn("can't find the file")
-      this.onClose()
-    }
-  }
+  const fetchFileIfNecessary = useCallback(
+    async isMounted => {
+      if (getCurrentIndex() !== -1) return
+      if (currentFile && isMounted) {
+        setCurrentFile(null)
+      }
 
-  async fetchMoreIfNecessary() {
-    if (this.fetchingMore) {
-      return
-    }
-    this.fetchingMore = true
-    try {
-      const { files, fileId, filesQuery } = this.props
-      const fileCount = filesQuery.count
-      // If we get close of the last file fetched, but we know there are more in the folder
-      // (it shouldn't happen in /recent), we fetch more files
-      const currentIndex = files.findIndex(f => f.id === fileId)
-
-      if (files.length !== fileCount && files.length - currentIndex <= 5) {
-        await filesQuery.fetchMore()
+      try {
+        const { data } = await client.query(Q('io.cozy.files').getById(fileId))
+        isMounted && setCurrentFile(data)
+      } catch (e) {
+        logger.warn("can't find the file")
+        handleOnClose()
+      }
+    },
+    [client, currentFile, fileId, getCurrentIndex, handleOnClose]
+  )
+
+  const fetchMoreIfNecessary = useCallback(
+    async isMounted => {
+      if (fetchingMore) return
+
+      setFetchingMore(true)
+      try {
+        const fileCount = filesQuery.count
+        // If we get close of the last file fetched, but we know there are more in the folder
+        // (it shouldn't happen in /recent), we fetch more files
+        const currentIndex = files.findIndex(f => f.id === fileId)
+
+        if (
+          files.length !== fileCount &&
+          files.length - currentIndex <= 5 &&
+          isMounted
+        ) {
+          await filesQuery.fetchMore()
+        }
+      } finally {
+        setFetchingMore(false)
       }
-    } finally {
-      this.fetchingMore = false
+    },
+    [fetchingMore, fileId, files, filesQuery]
+  )
+
+  useEffect(() => {
+    let isMounted = true
+
+    fetchFileIfNecessary(isMounted)
+    fetchMoreIfNecessary(isMounted)
```

Comment:
nit: Après coup je pense que le contenu de chacune des ces fonctions devrait être dans un useEffect différent.
Je ne suis pas sûr que l'astuce du `isMounted` soit fonctionnelle ainsi.

Reply:
Ton commentaire m'a mis la puce à l'oreille, qui n'est pas un `nit` je pense ;)

J'ai re-vérifié ce qu'on avait avec le class component, et dans le cas présent on a perdu le `componentWillReceiveProps() { this.fetchMoreIfNecessary() }` qu'on avait précédemment.

J'ai changé un peu l'approche, en effectivement mettant `fetchFileIfNecessary` et `fetchMoreIfNecessary` dans 2 `useEffect` séparé. J'ai aussi supprimé l'extraction de la fonction (qui avait un `useCallback`) pour la créer directement dans le `useEffect`, afin de mieux gérer les dépendances et rerender.

J'ai testé sur la page `/recent` et ça fonctionne bien, on n'a pas 36 rerender intempestif. De plus pour info on a un test qui vérifie `fetchFileIfNecessary` (donc qu'on fetch bien le premier fichier si y'a un souci) et `fetchMoreIfNecessary` (qui déclenche bien le `fetchMore()` quand on arrive à la fin de viewer)